azure-iot-device/azure/iot/device/common/transport/mqtt/pipeline_stages.py

```diff
@@ -0,0 +1,645 @@
+        """
+        logger.info("{}({}): completing".format(self.name, op.name))
+        op.callback(op)
+
+    def continue_with_translated_op(self, original_op, translated_op):
+        """
+        Continue an operation using a translated operation that carries on the work
+        represented by the operation.  When the translated operation completes, its
+        callback will cause the original operation to complete with the same result.
+
+        This is useful when a generic operation (such as "enable feature" needs to be
+        converted into a more specific operation (such as "subscribe to mqtt topic").
+        In that case, a stage's _run_op function would call this function passing in
+        the original "enable feature" op and the translated "subscribe to mqtt topic"
+        op.  This function will pass the subscribe op down, and when the subscribe op
+        is completed, this function will cause the original op to complete.
+
+        This function is only really useful if there is no data returned in the
+        translated_op that might need to be copied back into the original_op before
+        completing it.  If data needs to be copied this way, some other method needs
+        to be used.  (or a "copy data back" function needs to be added to this function
+        as an optional parameter.)
+        """
+
+        def translated_op_complete(op):
+            logger.info(
+                "%s: %s complete-returning result in %s",
+                self.name,
+                translated_op.name,
+                original_op.name,
+            )
+            original_op.error = translated_op.error
+            self.complete_op(original_op)
+
+        translated_op.callback = translated_op_complete
+        self.continue_op(translated_op)
+
+    # TODO: connected and disconnect migth become pipeline events.  maybe.
+    def on_connected(self):
+        """
+        Called by lower layers when the transport connects
+        """
+        if self.previous:
+            self.previous.on_connected()
+
+    def on_disconnected(self):
+        """
+        Called by lower layers when the transport disconnects
+        """
+        if self.previous:
+            self.previous.on_disconnected()
+
+
+class PipelineRoot(PipelineStage):
+    """
+    Object representing the root of a pipeline.  This is where the functions to build
+    the pipeline exist.  This is also where clients can add event handlers to receive
+    events from the pipeline.
+    """
+
+    def __init__(self):
+        PipelineStage.__init__(self)
+        self.on_pipeline_event = None
+
+    def _run_op(self, op):
+        """
+        run the operation.  At the root, the only thing to do is to pass the operation
+        to the next stage.
+        """
+        logger.info("{}({}) handling".format(self.name, op.name))
+        self.continue_op(op)
+
+    def add_next_stage(self, new_next_stage):
```

Comment:
>add_next_stage [](start = 8, length = 14)

terminology nit: couldn't we just call this 'append'? That term implies a lot of the detail of this, and you could then remove all references the verbose "new_next_stage". Plus "append" is the term used in the list primitive, so it's familiar